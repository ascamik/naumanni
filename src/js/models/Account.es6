import {Record} from 'immutable'
import {REGEX_PGP_FINGERPRINT} from 'src/constants'

const AccountRecord = Record({  // eslint-disable-line new-cap
  host: '',

  id: null,
  username: null,
  acct: null,
  display_name: null,
  locked: null,
  created_at: null,
  followers_count: null,
  following_count: null,
  statuses_count: null,
  note: null,
  url: null,
  avatar: null,
  avatar_static: null,
  header: null,
  header_static: null,
})


/**
 * OAuthでのApp登録を表すModel
 */
export default class Account extends AccountRecord {
  static storeName = 'accounts'
  static keyPath = 'address'
  static indexes = [
    ['host', 'host', {}],
  ]

  /**
   * @constructor
   */
  constructor(...args) {
    super(...args)
  }

  /**
   * @override
   */
  toJS() {
    const js = super.toJS()

    js.address = this.address
    return js
  }

  /**
   * 一意な識別子を返す
   */
  get address() {
    return `${this.id}@${this.host}`
  }

  get account() {
    if(this.acct.indexOf('@') >= 0)
      return this.acct
    return `${this.acct}@${this.host}`
  }

  get hasPublicKey() {
    return this.note.match(REGEX_PGP_FINGERPRINT) ? true : false
  }

  get publicKeyId() {
    const match = this.note.match(REGEX_PGP_FINGERPRINT)
    if(!match)
      return null
    const fingerprint = match[1]
    if(fingerprint.length != 40)
      return null
    return fingerprint.substring(24)
  }

  get privateKeyArmored() {
    return localStorage.getItem(`pgp::privateKey::${this.address}`)
  }
}
