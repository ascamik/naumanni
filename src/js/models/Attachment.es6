import {Record} from 'immutable'


const AttachmentRecord = Record({  // eslint-disable-line new-cap
  'id': '',
  'type': '',
  'url': '',
  'remote_url': '',
  'preview_url': '',
  'text_url': '',
})


/**
 * Mastodon„ÅÆMedia
 */
export default class Attachment extends AttachmentRecord {
}
