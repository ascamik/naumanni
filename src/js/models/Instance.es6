/* @flow */
import {fromJS, Map, Record} from 'immutable'

import type OAuthToken from 'src/models/OAuthToken'

import {
  STREAM_HOME, STREAM_LOCAL, STREAM_FEDERATION, STREAM_TAG,
} from 'src/constants'

type Url = string
type InstanceShape = {
  uri: string,
  title: string,
  description: string,
  email: string,
  version: string,
  urls: Map<string, string>
};


const URL_KEY_STREAMING_API = 'streaming_api'
const instanceDefaults: InstanceShape = {
  host: '',
  uri: '',
  title: '',
  description: '',
  email: '',
  version: '',
  urls: new Map(),
}

/**
 * Mastodon Instance Information
 * see https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md#instance
 */
export default class Instance extends Record(instanceDefaults) {  // eslint-disable-line new-cap
  /**
   * @constructor
   * @param {string} host
   * @param {any} raw
   */
  constructor(host: string, raw: {urls?: {[string]: string}}) {
    super({
      ...raw,
      host,
      urls: fromJS(raw.urls || {}),
    })
  }

  /**
   * WSSのURLを返す
   * @param {OAuthToken} token
   * @param {string} stream
   * @param {object} params
   * @return {Url}
   */
  makeStreamingAPIUrl(token: OAuthToken, stream: string, params?: any={}): Url {
    require('assert')(
      [STREAM_FEDERATION, STREAM_TAG, STREAM_LOCAL, STREAM_HOME].indexOf(stream) >= 0,
      'valid stream type'
    )

    let rootUrl = this.urls.get(URL_KEY_STREAMING_API)
    if(!rootUrl) {
      rootUrl = `wss://${token.host}`
    }

    const otherParams = Object.keys(params).length === 0
      ? ''
      : Object.keys(params)
          .reduce((prev, k, i) => (
            i === 0
              ? `&${k}=${encodeURIComponent(params[k])}`
              : `${prev}&${k}=${encodeURIComponent(params[k])}`
          ), '')

    return `${rootUrl}/api/v1/streaming/?access_token=${token.accessToken}&stream=${stream}${otherParams}`
  }
}
