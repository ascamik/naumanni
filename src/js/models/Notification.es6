import {Record} from 'immutable'
import moment from 'moment'


const NotificationRecord = Record({  // eslint-disable-line new-cap
  host: '',

  id: null,
  type: '',
  created_at: '',
  account: null,
  status: null,
})


/**
 * Mastodon's Notification
 */
export default class Notification extends NotificationRecord {
  get uri() {
    return `notification:${this.host}:${this.id}`
  }

  get createdAt() {
    return moment(this.created_at)
  }

  // TODO: どっかに纏める
  static compareCreatedAt(a, b) {
    const aAt = a.createdAt
    const bAt = b.createdAt
    if(aAt.isBefore(bAt))
      return 1
    else if(aAt.isAfter(bAt))
      return -1
    return 0
  }
}
