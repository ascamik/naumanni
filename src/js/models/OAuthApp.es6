import {Record} from 'immutable'


const OAuthAppRecord = Record({  // eslint-disable-line new-cap
  host: '',
  id: '',
  redirect_uri: '',
  client_id: '',
  client_secret: '',
})


/**
 * OAuthでのApp登録を表すModel
 */
export default class OAuthApp extends OAuthAppRecord {
  static storeName = 'oauth_apps'
  static keyPath = 'address'
  static indexes = [
    ['host', 'host', {}],  // uniqueでも良いかも
    ['client_id', ['host', 'client_id'], {unique: true}],
  ]

  /**
   * @constructor
   */
  constructor(...args) {
    super(...args)
  }

  /**
   * @override
   */
  toJS() {
    const js = super.toJS()

    js.address = this.address
    return js
  }

  /**
   * 一意な識別子を返す
   */
  get address() {
    return `${this.host}::${this.id}`
  }
}
