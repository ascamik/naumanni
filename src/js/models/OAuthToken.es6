import moment from 'moment'
import {Record} from 'immutable'

import {makeOAuthAPIRequester} from 'src/api/APIRequester'
import MastodonAPISpec from 'src/api/MastodonAPISpec'


const OAuthTokenRecord = Record({  // eslint-disable-line new-cap
  host: '',

  access_token: null,
  refresh_tokend: null,
  created_at: null,
  scope: '',
  token_type: '',
})


/**
 * OAuthTokenを表す
 */
export default class OAuthToken extends OAuthTokenRecord {
  static storeName = 'oauth_tokens'
  static keyPath = 'address'
  static indexes = [
    ['access_token', ['host', 'access_token'], {unique: true}],
  ]

  /**
   * @constructor
   * @param {Objet} token
   */
  constructor(...args) {
    super(...args)
  }

  toJS() {
    const js = super.toJS()

    js.address = this.address
    return js
  }

  get accessToken() {
    return this.access_token
  }

  get address() {
    return `${this.host}::${this.access_token}`
  }

  get expiresAt() {
    // Mastodonでは使わん
    // const {
    //   access_token,
    //   refresh_token,
    //   expires,
    //   expires_in,
    // } = token

    // require('assert')(access_token && refresh_token)
    // this.accessToken = access_token
    // this.refreshToken = refresh_token

    // require('assert')(!(expires && expires_in), 'expires and expires_in cannot set simultaneously')
    // require('assert')(expires || expires_in, 'expires or expires_in is required')

    // if(expires_in) {
    //   // 時刻設定連れを考慮して10%の誤差を...
    //   this.expires = moment().add(expires_in * 0.9, 's')
    // } else {
    //   this.expires = expires
    // }
    throw new Error('not implemented')
  }

  /**
   * APIRequesterを作る
   * TODO: 場所を考える
   */
  get requester() {
    if(!this._requester) {
      this._requester = makeOAuthAPIRequester(
        MastodonAPISpec, {
          token: this,
          endpoint: `https://${this.host}/api/v1`,
          hooks: {
            response: (method, apiName, responseBody) => {
              return {host: this.host, ...responseBody}
            },
          },
        })
    }
    return this._requester
  }

  /**
   * TokenがexpireしていたらTrueを返す
   * @return {bool}
   */
  isExpired() {
    // refresh_tokenが無い場合はexpiresしないと考える?
    if(!this.refresh_token)
      return false

    return this.expires.isBefore(moment())
  }
}
