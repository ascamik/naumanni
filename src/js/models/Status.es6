import moment from 'moment'
import {Record} from 'immutable'

import {
  MESSAGE_TAG_REX,
  VISIBLITY_DIRECT, VISIBLITY_PRIVATE, VISIBLITY_UNLISTED, VISIBLITY_PUBLIC,
} from 'src/constants'


const StatusRecord = Record({  // eslint-disable-line new-cap
  host: '',

  id: '',
  uri: '',
  url: '',
  account: null,
  in_reply_to_id: '',
  in_reply_to_account_id: '',
  reblog: null,
  content: '',
  created_at: '',
  reblogs_count: '',
  favourites_count: '',
  reblogged: '',
  favourited: '',
  sensitive: '',
  spoiler_text: '',
  visibility: '',
  media_attachments: [],
  mentions: '',
  tags: '',
  application: '',
})


/**
 * MastodonのStatus
 */
export default class Status extends StatusRecord {
  /**
   * @constructor
   * @param {object} raw
   */
  constructor(raw) {
    if(raw.account) {
      // avoid circular dependency
      const Account = require('./Account').default
      raw.account = new Account({
        host: raw.host,
        ...raw.account,
      })
    }
    if(raw.reblog) {
      raw.reblog = new Status({
        host: raw.host,
        ...raw.reblog,
      })
    }

    if(raw.media_attachments.length) {
      const Attachment = require('./Attachment').default
      raw.media_attachments = raw.media_attachments.map((rawmedia) => new Attachment(rawmedia))
    }

    super(raw)
  }

  get rawContent() {
    // TODO:sanitize
    // import {sanitizeHtml} from 'sanitize-html'
    return this.content
  }

  get createdAt() {
    return moment(this.created_at)
  }

  get hasSpoilerText() {
    return this.spoiler_text.length > 0
  }

  get spoilerText() {
    return this.spoiler_text
  }

  canReblog() {
    return (this.visibility === VISIBLITY_PUBLIC || this.visibility === VISIBLITY_UNLISTED)
      ? true
      : false
  }

  /**
   * そいつあてのMentionが含まれているか？
   * @param {Int} accountId そいつ
   * @return {bool}
   */
  isMentionToId(accountId) {
    return this.mentions.find((m) => m.id === accountId) ? true : false
  }

  static compareCreatedAt(a, b) {
    const aAt = a.created_at
    const bAt = b.created_at
    if(aAt < bAt)
      return 1
    else if(aAt > bAt)
      return -1
    return 0
  }

  // naumanni用機能
  get messageBlockInfo() {
    const match = this.content.match(MESSAGE_TAG_REX)
    if(!match)
      return null

    return {
      checksum: match[1],
      index: +match[2],
      total: +match[3],
    }
  }
}
