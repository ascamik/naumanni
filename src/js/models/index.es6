import Account from './Account'
import Attachment from './Attachment'
import OAuthApp from './OAuthApp'
import OAuthToken from './OAuthToken'
import Status from './Status'
export {Account, OAuthApp, OAuthToken, Status}

import UIColumn from './UIColumn'
export {UIColumn}
