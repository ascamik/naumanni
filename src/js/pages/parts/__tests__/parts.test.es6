import React from 'react';
import renderer from 'react-test-renderer';

import {Account} from 'src/models'
import {DropdownMenuButton, IconFont, NowLoading, UserIconWithHost} from '../'


describe('DropdownMenuButton', () => {
  it('can render button', () => {
    const tree = renderer.create(
      <DropdownMenuButton onRenderMenu={() => <div>menu!!</div>}>
        <div>child!!</div>
      </DropdownMenuButton>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('can show and hide menu', () => {
    const component = renderer.create(
      <DropdownMenuButton onRenderMenu={() => <div><a href="http://goo.gl">NiceLink</a></div>}>
        <div>child!!</div>
      </DropdownMenuButton>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot('DropdownMenuButton can show and hide menu : only button is visible')

    // menu was showed
    expect(tree.children[0].props.className).toBe('dropdownMenuButton-button')
    tree.children[0].props.onClick({preventDefault: jest.fn(), stopPropagation: jest.fn()})
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('DropdownMenuButton can show and hide menu : button and menu are visible')

    // menu was hidden
    tree.props.onMouseLeave()
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('DropdownMenuButton can show and hide menu : only button is visible')

    // menu was showed
    tree.children[0].props.onClick({preventDefault: jest.fn(), stopPropagation: jest.fn()})
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('DropdownMenuButton can show and hide menu : button and menu are visible')

    // menu was closed after link click
    expect(tree.children[1].props.className).toBe('dropdownMenuButton-menu')
    tree.children[1].props.onClick({preventDefault: jest.fn(), stopPropagation: jest.fn()})
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('DropdownMenuButton can show and hide menu : only button is visible')
  })
})


test('IconFont', () => {
  const tree = renderer.create(
    <IconFont iconName="iconA" />
  ).toJSON()
  expect(tree).toMatchSnapshot()

  const treeB = renderer.create(
    <IconFont iconName="iconB" />
  ).toJSON()
  expect(treeB).toMatchSnapshot()
})


test('NowLoading', () => {
  const tree = renderer.create(
    <NowLoading />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})


describe('UserIconWithHost', () => {
  it('can show user icon', () => {
    const mockAccount = new Account({
      acct: 'shn@oppai.tokyo',
      avatar: 'https://ot-mastodon.s3.amazonaws.com/accounts/avatars/000/000/001/original/2408e330e310f168.png?1492583237',
      display_name: 'Shin Adachi',
    })
    const tree = renderer.create(
      <UserIconWithHost account={mockAccount} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('can show default icon if not avatar was present', () => {
    const mockAccount = new Account({
      acct: 'shn@oppai.tokyo',
      avatar: null,
      display_name: 'Shin Adachi',
    })
    const tree = renderer.create(
      <UserIconWithHost account={mockAccount} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
