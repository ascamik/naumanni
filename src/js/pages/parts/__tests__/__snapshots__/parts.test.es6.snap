// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenuButton can render button 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton can show and hide menu : button and menu are visible 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
  <div
    className="dropdownMenuButton-menu"
    onClick={[Function]}
  >
    <div>
      <a
        href="http://goo.gl"
      >
        NiceLink
      </a>
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton can show and hide menu : button and menu are visible 2`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
  <div
    className="dropdownMenuButton-menu"
    onClick={[Function]}
  >
    <div>
      <a
        href="http://goo.gl"
      >
        NiceLink
      </a>
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton can show and hide menu : only button is visible 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton can show and hide menu : only button is visible 2`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton can show and hide menu : only button is visible 3`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <div>
      child!!
    </div>
  </div>
</div>
`;

exports[`IconFont 1`] = `
<span
  className="icon icon-iconA"
/>
`;

exports[`IconFont 2`] = `
<span
  className="icon icon-iconB"
/>
`;

exports[`NowLoading 1`] = `
<div
  className="niceLoader"
/>
`;

exports[`UserIconWithHost can show default icon if not avatar was present 1`] = `
<span
  className="userIcon with-host"
  onClick={undefined}
>
  <img
    alt="shn@oppai.tokyo"
    className="userIcon-avatar"
    src="/images/no-avatar.png"
    title="shn@oppai.tokyo"
  />
  <img
    className="userIcon-host"
    src="https://oppai.tokyo/favicon.ico"
  />
</span>
`;

exports[`UserIconWithHost can show user icon 1`] = `
<span
  className="userIcon with-host"
  onClick={undefined}
>
  <img
    alt="shn@oppai.tokyo"
    className="userIcon-avatar"
    src="https://ot-mastodon.s3.amazonaws.com/accounts/avatars/000/000/001/original/2408e330e310f168.png?1492583237"
    title="shn@oppai.tokyo"
  />
  <img
    className="userIcon-host"
    src="https://oppai.tokyo/favicon.ico"
  />
</span>
`;
